### DISCUSSION TITLE
Error building on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by jimsyt** [abraunegg/onedrive#2218] [Q&A] [answered]
i get the following errors when trying to build on ubuntu 20.04:
```
if [ -f .git/HEAD ] ; then \
	git describe --tags > version ; \
else \
	echo v2.4.22-dev > version ; \
fi
dmd  -w -g -O -J. -version=NoPragma -version=NoGdk -version=Notifications -L-lcurl -L-lsqlite3 -L-lnotify -L-lgdk_pixbuf-2.0 -L-lgio-2.0 -L-lgobject-2.0 -L-lglib-2.0 -L-ldl src/config.d src/itemdb.d src/log.d src/main.d src/monitor.d src/onedrive.d src/qxor.d src/selective.d src/sqlite.d src/sync.d src/upload.d src/util.d src/progress.d src/arsd/cgi.d src/notifications/notify.d src/notifications/dnotify.d -ofonedrive
src/arsd/cgi.d(4730): Error: function `@trusted long arsd.cgi.FakeSocketForStdin.receive(void[] buffer, SocketFlags)` does not override any function, did you mean to override `@trusted long std.socket.Socket.receive(scope void[] buf, std.socket.SocketFlags flags)`?
src/arsd/cgi.d(4736): Error: function `@trusted long arsd.cgi.FakeSocketForStdin.send(const(void[]) buffer, SocketFlags)` does not override any function, did you mean to override `@trusted long std.socket.Socket.send(scope const(void)[] buf, std.socket.SocketFlags flags)`?
src/arsd/cgi.d(4743): Error: function `@trusted void arsd.cgi.FakeSocketForStdin.close()` does not override any function, did you mean to override `nothrow @nogc scope @trusted void std.socket.Socket.close()`?
src/arsd/cgi.d(4751): Error: function `void arsd.cgi.FakeSocketForStdin.setOption(SocketOptionLevel, SocketOption, void[])` does not override any function, did you mean to override `@trusted void std.socket.Socket.setOption(std.socket.SocketOptionLevel level, std.socket.SocketOption option, scope void[] value)`?
make: *** [Makefile:102: onedrive] Error 1
```

I'm using (dmd-2.101.0)

any ideas?

**Comment by abraunegg** [comment, marked as answer]
@jimsyt 
> any ideas?

It appears that with the updated DMD, with the latest version breaks an included file - so this will need to be looked at as to if there is an updated upstream `cgi.d` or if the current one can be modified as required to work with the latest DMD compiler.

Your options for the moment:
1. As you are using Ubuntu, it is *highly* recommended that you use the OpenSuSE Build Service packages. Please read: https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md
2. If you *really* want to compile, I have tested '2.100.2' and this does not have the `cgi.d` issue you have found. You can install using the direct version command: `curl -fsS https://dlang.org/install.sh | bash -s dmd-2.100.2`


**Edit|Update:**
This issue|problem is now resolved in 'master' with https://github.com/abraunegg/onedrive/pull/2219 

```
(dmd-2.101.0)[alex@onedrive-client-dev onedrive]$ make clean; make;
rm -f onedrive onedrive.o version
rm -rf autom4te.cache
rm -f config.log config.status
if [ -f .git/HEAD ] ; then \
        git describe --tags > version ; \
else \
        echo v2.4.22-dev > version ; \
fi
dmd -w -g -O -J. -L-lcurl -L-lsqlite3  -L-ldl src/config.d src/itemdb.d src/log.d src/main.d src/monitor.d src/onedrive.d src/qxor.d src/selective.d src/sqlite.d src/sync.d src/upload.d src/util.d src/progress.d src/arsd/cgi.d -ofonedrive
(dmd-2.101.0)[alex@onedrive-client-dev onedrive]$ 
```

**Reply by jimsyt** [reply]
I did end up installing using the OpenSuSE method but that seemed kind of odd to me to begin with.

update does seem to work perfectly as well - Thanks!

**Reply by abraunegg** [reply]
@jimsyt 
> I did end up installing using the OpenSuSE method but that seemed kind of odd to me to begin with

Why is this odd to you? Can you explain please.

**Reply by jimsyt** [reply]
I guess I didn't really read it to begin with. I thought it was going to be some way to convert an OpenSuSE package over to install on ubuntu but looking at it again, it looks like its just an OpenSuSE repo that holds ubuntu packages?

**Reply by abraunegg** [reply]
> ... it looks like its just an OpenSuSE repo that holds ubuntu packages?

Correct. Due to the issues with package management and package maintenance on Ubuntu, it is easier to have the packages built and hosted on the OpenSuSE Build Service platform.
